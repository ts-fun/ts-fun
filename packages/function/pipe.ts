export { pipe } from 'ts-pipe-compose'
