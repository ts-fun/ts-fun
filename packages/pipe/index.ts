export * from '@tsfun/tap'

class PipeNode<Args extends any[], Return> {
  /**
   * Resulting function of the pipeline
   *
   * @param args It's either
   *   * the same parameters of the first function of the pipeline (when `pipe` is used)
   *   * or empty (when `pass` is used)
   *
   * @returns Result of the last function of the pipeline
   */
  public readonly get: (...args: Args) => Return

  constructor(fn: (...args: Args) => Return) {
    this.get = fn
  }

  /**
   * Add a node to the pipeline
   * @param fn Function of the node
   * @param tailArgs Parameters of `fn` without the first one
   * @returns Another node of the pipeline
   */
  public to<NextReturn, Tail extends any[]>(
    fn: (pipedValue: Return, ...tailArgs: Tail) => NextReturn,
    ...tailArgs: Tail
  ) {
    return pipe(
      (...firstArgs: Args) => fn(this.get(...firstArgs), ...tailArgs),
    )
  }
}

/**
 * Start a pipeline with a value
 * (pass a value to a series of functions in a pipeline)
 *
 * **Example:**
 *
 * ```typescript
 * const increase = (x: number) => x + 1
 * const double = (x: number) => x * 2
 * const square = (x: number) => x * x
 * const result = pass(2) // x0 = 2
 *   .to(increase) // x1 = increase(x0)
 *   .to(double) // x2 = double(x1)
 *   .to(square) // x3 = square(x2)
 *   .to(String) // x4 = String(x3)
 *   .get() // result = x4 = String(square(double(increase(2))))
 * expect(result).toBe(String(square(double(increase(2)))))
 * ```
 *
 * **Example:**
 *
 * ```typescript
 * const sum = (...args: number[]) => args.reduce((a, b) => a + b)
 * const product = (...args: number[]) => args.reduce((a, b) => a * b)
 * const result = pass(2) // x0 = 2
 *   .to(sum, -3, 4) // x1 = sum(x0, -3, 4)
 *   .to(product, 2) // x2 = product(x1, 2)
 *   .to(String) // x3 = String(x2)
 *   .get() // result = x3 = String(product(sum(2, -3, 4), 2))
 * expect(result).toBe(String(product(sum(2, -3, 4), 2)))
 * ```
 *
 * **Usage:**
 *
 * `pass(x).to(f).to(g).to(h).get()` is equivalent to `x |> f |> g |> h`,
 * which in turn is equivalent to `h(g(f(x)))`
 *
 * You can also pass extra arguments to function that takes more than one argument,
 * for example: `pass(x).to(f, y, z).get()` is equivalent to `f(x, y, z)`
 *
 * **Note:**
 *
 * This is a temporary solution to replace yet-to-be-implemented pipeline operator.
 * This function will be deprecated when the proposal get to stage 3
 * and removed when it is implemented
 *
 * Read more about pipeline operator in
 * [the tc39 proposal](https://github.com/tc39/proposal-pipeline-operator)
 *
 * @param x First value in the pipeline
 * @returns First node of the pipeline
 */
export const pass = <X>(x: X) => pipe(() => x)

/**
 * Start a pipeline with a function
 *
 * **Example:**
 *
 * ```typescript
 * const sum = (...args: number[]) => args.reduce((a, b) => a + b, 0)
 * const increase = (x: number) => x + 1
 * const double = (x: number) => x * 2
 * const square = (x: number) => x * x
 * const pipeline = pipe(sum) // x0 = sum(...args)
 *   .to(increase) // x1 = increase(x0)
 *   .to(double) // x2 = double(x1)
 *   .to(square) // x3 = square(x2)
 *   .to(String) // x4 = String(x3)
 *   .get // pipeline(...args) = x4 = String(square(double(increase(sum(...args)))))
 * expect(pipeline(2, 3, -1)).toBe(String(square(double(increase(sum(2, 3, -1))))))
 * ```
 *
 * **Example:**
 *
 * ```typescript
 * const sum = (...args: number[]) => args.reduce((a, b) => a + b)
 * const product = (...args: number[]) => args.reduce((a, b) => a * b)
 * const pipeline = pipe(sum) // x0 = sum(...args)
 *   .to(product, 2) // x1 = product(x0, 2)
 *   .to(String) // x2 = String(x1)
 *   .get // pipeline(...args) = x2 = String(product(sum(2, -3, 4), 2))
 * expect(pipeline(2, -3, 4)).toBe(String(product(sum(2, -3, 4), 2)))
 * ```
 *
 * **Usage:**
 * `pipe(f).to(g).to(h).get(...args)` is equivalent to `f(...args) |> g |> h`,
 * which in turn is equivalent to `h(g(f(...args)))`
 *
 * You can also pass extra arguments to function that takes more than takes more than one arguments,
 * for example: `pipe(f).to(g, a, b).get` is equivalent to `(...args) => g(f(...args), a, b)`
 *
 * **Note:**
 * This is a temporary solution to replace yet-to-be-implemented pipeline operator.
 * This function will be deprecated when the proposal get to stage 3
 * and removed when it is implemented
 *
 * Read more about pipeline operator in
 * [the tc39 proposal](https://github.com/tc39/proposal-pipeline-operator)
 *
 * @param fn First function in the pipeline
 * @returns First node of the pipeline
 */
export const pipe = <Args extends any[], Return>(fn: (...args: Args) => Return) => new PipeNode(fn)
