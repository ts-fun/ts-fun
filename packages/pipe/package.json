{
  "name": "@tsfun/pipe",
  "version": "0.0.7",
  "description": "Dot-chained pipeline",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://tsfun.github.io/",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/tsfun/tsfun.git"
  },
  "bugs": {
    "url": "https://github.com/tsfun/tsfun/issues"
  },
  "keywords": [
    "pipe",
    "pipeline"
  ],
  "sideEffects": false,
  "dependencies": {
    "tslib": "^1.10.0",
    "@tsfun/tap": "^0.0.6",
    "@types/node": "^12.7.11"
  },
  "engines": {
    "node": ">= 8.9.0"
  }
}
