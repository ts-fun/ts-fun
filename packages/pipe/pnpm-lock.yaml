dependencies:
  '@tsfun/tap': 'link:../tap'
  '@types/node': 13.7.4
  tslib: 1.11.0
lockfileVersion: 5.1
packages:
  /@types/node/13.7.4:
    dev: false
    resolution:
      integrity: sha512-oVeL12C6gQS/GAExndigSaLxTrKpQPxewx9bOcwfvJiJge4rr7wNaph4J+ns5hrmIV2as5qxqN8YKthn9qh0jw==
  /tslib/1.11.0:
    dev: false
    resolution:
      integrity: sha512-BmndXUtiTn/VDDrJzQE7Mm22Ix3PxgLltW9bSNLoeCY31gnG2OPx0QqJnuc9oMIKioYrz487i6K9o4Pdn0j+Kg==
specifiers:
  '@tsfun/tap': ^0.0.10
  '@types/node': ^13.7.4
  tslib: ^1.11.0
