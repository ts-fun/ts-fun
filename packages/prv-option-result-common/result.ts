import { Ok, Err } from './types'
import { box } from './box'

/**
 * Create an `Ok`
 * @param payload Payload to carry
 * @returns `Ok` of `payload`
 */
export const ok: <Payload> (payload: Payload) => Ok<Payload> = box

/**
 * Create an `Err`
 * @param error Error to carry
 * @returns `Err` of `error`
 */
export const err = <Error> (error: Error): Err<Error> => ({ tag: false, error })
