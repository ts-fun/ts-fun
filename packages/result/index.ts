export * from './result'
export * from './ok'
export * from './err'
export * from './expect-some'
export * from './and'
export * from './or'
export * from './flatten'
export * from './async-flatten'
export * from './unwrap'
export * from './async-unwrap'
export * from './clone'
export * from './flip'
export * from './match'
export * from './map'
export * from './transpose'
export * from './try-exec'
export * from './tap'
