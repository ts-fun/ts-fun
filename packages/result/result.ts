export { Result } from '@ts-fun/prv-option-result-common'
