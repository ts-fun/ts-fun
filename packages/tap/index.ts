export interface SideEffect<Value> {
  (value: Value): void
}

/**
 * Insert a side-effect into function pipeline
 * @param value Value from pipeline
 * @param fns Side effect functions
 * @returns `value`
 *
 * @example
 *   const result = pass(value)
 *     .to(tap, value => console.log('begin', value))
 *     .to(inc)
 *     .to(tap, value => console.log('after inc', value))
 *     .to(double)
 *     .to(tap, value => console.log('after double', value))
 *     .to(square)
 *     .to(tap, value => console.log('after square', value))
 *     .get()
 */
export function tap<Value> (value: Value, ...fns: SideEffect<Value>[]): Value {
  fns.forEach(fn => fn(value))
  return value
}

export interface Tapper<Value> {
  <Sub extends Value> (value: Sub): Sub
}

/**
 * Insert a side-effect into function pipeline.
 * This is used in situation where `tap` is not suited,
 * such as `map`
 * @param fns Side effect functions
 * @returns A side-effect function that can be inserted into a pipeline
 *
 * @example
 *   const myResult = pass(result)
 *     .to(tap, result => console.log('result before onOk/onErr', result))
 *     .to(map, tapper(value => console.log('value before onOk', value)))
 *     .to(mapErr, tapper(error => console.log('error before onErr', error)))
 *     .to(map, onOk)
 *     .to(map, tapper(value => console.log('value after onOk', value)))
 *     .to(mapErr, onErr)
 *     .to(mapErr, tapper(error => console.log('error after onErr', error)))
 *     .get()
 */
export function tapper<Value> (...fns: SideEffect<Value>[]): Tapper<Value> {
  return value => tap(value, ...fns)
}
