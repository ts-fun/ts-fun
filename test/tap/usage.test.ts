import { pass } from '@tsfun/pipe'
import { Result, ok, err, map, mapErr } from '@tsfun/result'
import { tap, tapper } from '@tsfun/tap'

describe('tap: @tsfun/pipe → pass → logging', () => {
  const setup = () => {
    const logs = Array<any>()
    const logger = (id: any) => (value: any) => logs.push({ id, value })
    const result = pass(2)
      .to(tap, logger('begin'))
      .to(x => x + 1)
      .to(tap, logger('inc'))
      .to(x => x * 2)
      .to(tap, logger('double'))
      .to(x => x * x)
      .to(tap, logger('square'))
      .get()
    return { logs, result }
  }

  it('produces logs that matches snapshot', () => {
    const { logs } = setup()
    expect(logs).toMatchSnapshot()
  })

  it('returns value that matches snapshot', () => {
    const { result } = setup()
    expect(result).toMatchSnapshot()
  })
})

describe('tapper: @tsfun/result → map + mapErr → logging', () => {
  const setup = (input: Result<'ok', 'err'>) => {
    const logs = Array<any>()
    const logger = (id: any, key: string) => (value: any) => logs.push({ id, [key]: value })
    const output = pass(input)
      .to(map, tapper(logger('before', 'input.value')))
      .to(mapErr, tapper(logger('before', 'input.error')))
      .to(map, (x: 'ok') => x.toUpperCase())
      .to(map, tapper(logger('after', 'input.value')))
      .to(mapErr, (x: 'err') => x.toUpperCase())
      .to(mapErr, tapper(logger('after', 'input.error')))
      .get()
    return { logs, output }
  }

  describe('ok("ok")', () => {
    it('produces logs that matches snapshot', () => {
      const { logs } = setup(ok('ok'))
      expect(logs).toMatchSnapshot()
    })

    it('returns value that matches snapshot', () => {
      const { output } = setup(ok('ok'))
      expect(output).toMatchSnapshot()
    })
  })

  describe('err("err")', () => {
    it('produces logs that matches snapshot', () => {
      const { logs } = setup(err('err'))
      expect(logs).toMatchSnapshot()
    })

    it('returns value that matches snapshot', () => {
      const { output } = setup(err('err'))
      expect(output).toMatchSnapshot()
    })
  })
})
