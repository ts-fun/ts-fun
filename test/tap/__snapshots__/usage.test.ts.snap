// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tap: @tsfun/pipe → pass → logging produces logs that matches snapshot 1`] = `
Array [
  Object {
    "id": "begin",
    "value": 2,
  },
  Object {
    "id": "inc",
    "value": 3,
  },
  Object {
    "id": "double",
    "value": 6,
  },
  Object {
    "id": "square",
    "value": 36,
  },
]
`;

exports[`tap: @tsfun/pipe → pass → logging returns value that matches snapshot 1`] = `36`;

exports[`tapper: @tsfun/result → map + mapErr → logging err("err") produces logs that matches snapshot 1`] = `
Array [
  Object {
    "id": "before",
    "input.error": "err",
  },
  Object {
    "id": "after",
    "input.error": "ERR",
  },
]
`;

exports[`tapper: @tsfun/result → map + mapErr → logging err("err") returns value that matches snapshot 1`] = `
Object {
  "error": "ERR",
  "tag": false,
}
`;

exports[`tapper: @tsfun/result → map + mapErr → logging ok("ok") produces logs that matches snapshot 1`] = `
Array [
  Object {
    "id": "before",
    "input.value": "ok",
  },
  Object {
    "id": "after",
    "input.value": "OK",
  },
]
`;

exports[`tapper: @tsfun/result → map + mapErr → logging ok("ok") returns value that matches snapshot 1`] = `
Object {
  "tag": true,
  "value": "OK",
}
`;
